File: extr_api/internal/storage/postgres/operations.go
package postgres

import (
	"alex_gorbunov_exptr_api/internal/models"
	"fmt"
)

func (s *Storage) CreateOperation(operation models.CreateOperationRequest) error {
	const fn = "storage.postgresql.CreateOperation"

	query := `INSERT INTO operations (user_id, category_id, amount, currency, name, comment, type, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`
	_, err := s.db.Exec(query, operation.UserID, operation.CategoryID, operation.Amount, operation.Currency, operation.Name, operation.Comment, operation.Type, operation.CreatedAt, operation.UpdatedAt)
	if err != nil {
		return fmt.Errorf("%s: %w", fn, err)
	}

	return nil
}

func (s *Storage) UpdateOperation(operation *models.Operation) error {
	const fn = "storage.postgresql.UpdateOperation"

	query := `UPDATE operations SET user_id = $1, category_id = $2, amount = $3, currency = $4, name = $5, comment = $6, type = $7, updated_at = $8 WHERE id = $9`
	_, err := s.db.Exec(query, operation.UserID, operation.CategoryID, operation.Amount, operation.Currency, operation.Name, operation.Comment, operation.Type, operation.UpdatedAt, operation.ID)
	if err != nil {
		return fmt.Errorf("%s: %w", fn, err)
	}

	return nil
}

func (s *Storage) GetOperationByID(id string) (*models.Operation, error) {
	const fn = "storage.postgresql.GetOperationByID"

	query := `SELECT id, user_id, category_id, amount, currency, name, comment, type, created_at, updated_at FROM operations WHERE id = $1`
	row := s.db.QueryRow(query, id)

	var operation models.Operation
	err := row.Scan(&operation.ID, &operation.UserID, &operation.CategoryID, &operation.Amount, &operation.Currency, &operation.Name, &operation.Comment, &operation.Type, &operation.CreatedAt, &operation.UpdatedAt)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", fn, err)
	}

	return &operation, nil
}

func (s *Storage) GetOperationsByUserID(userID string) ([]models.Operation, error) {
	const fn = "storage.postgresql.GetOperationsByUserID"

	query := `SELECT id, user_id, category_id, amount, currency, name, comment, type, created_at, updated_at FROM operations WHERE user_id = $1`
	rows, err := s.db.Query(query, userID)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", fn, err)
	}
	defer rows.Close()

	var operations []models.Operation
	for rows.Next() {
		var operation models.Operation
		err := rows.Scan(&operation.ID, &operation.UserID, &operation.CategoryID, &operation.Amount, &operation.Currency, &operation.Name, &operation.Comment, &operation.Type, &operation.CreatedAt, &operation.UpdatedAt)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", fn, err)
		}
		operations = append(operations, operation)
	}

	return operations, nil
}

func (s *Storage) DeleteOperation(id string) error {
	const fn = "storage.postgresql.DeleteOperation"

	query := `DELETE FROM operations WHERE id = $1`
	_, err := s.db.Exec(query, id)
	if err != nil {
		return fmt.Errorf("%s: %w", fn, err)
	}

	return nil
}

File: exptr_api_p/exptr_api_p/repositories/operations.py
from sqlalchemy.orm import Session

from exptr_api_p.entities import operations
from exptr_api_p.repositories.database import operations as db_operations

class OperationsRepository:
  def __init__(self, session: Session):
    self.session = session

  def create_operation(self, operation: operations.OperationRequest):
    db_operation = db_operations.Operation(
      user_id=operation.user_id,
      category_id=operation.category_id,
      amount=operation.amount,
      currency=operation.currency,
      name=operation.name,
      comment=operation.comment,
      type=operation.type,
      created_at=operation.created_at,
      updated_at=operation.updated_at
    )
    self.session.add(db_operation)
    self.session.commit()
    self.session.refresh(db_operation)
  
  def update_operation(self, operation: operations.Operation):
    db_operation = self.session.query(db_operations.Operation).filter(db_operations.Operation.id == operation.id).first()
    db_operation.user_id = operation.user_id
    db_operation.category_id = operation.category_id
    db_operation.amount = operation.amount
    db_operation.currency = operation.currency
    db_operation.name = operation.name
    db_operation.comment = operation.comment
    db_operation.type = operation.type
    db_operation.updated_at = operation.updated_at
    self.session.commit()
    self.session.refresh(db_operation)
  
  def get_operation_by_id(self, operation_id: int):
    return self.session.query(db_operations.Operation).filter(db_operations.Operation.id == operation_id).first()
  
  def get_operations_by_user_id(self, user_id: int):
    return self.session.query(db_operations.Operation).filter(db_operations.Operation.user_id == user_id)
  
  def delete_operation(self, operation_id: int):
    self.session.query(db_operations.Operation).filter(db_operations.Operation.id == operation_id).delete()
    self.session.commit()